joomla/joomla-cms
# Implement a fallback for redirect urls

#### Introduction

The current implementation of the redirect application (component and plugin) expects a direct match to act uppon. So, if one wants to manually add a "rewrite rule" for e.g. http(s)://www.domain.tld/path/some-alias?some=param the rule would have to be exactly that url.

There are a couple of scenarios where query params could be added to any url, like when using a newsletter tool or if you've implemented certain affiliate mechanisms, where, especially in case of a relaunch of a website, you would have to create a whole bunch of redirect rules for each and every newsletter and / or affliiate link. Seldom this amount of different redirects can or should be handled via rewrite rules in a .htacces file.

This PR will add to the default behavior in a way that, if no specific (and published) redirect rule was found, it keeps looking for a "fallback".

Here is an example (behavior without the patch).:
If no (published) redirect rule for http(s)://www.domain.tld/path/some-alias?some=param exists, it will lead to a 404 status code.

After the patch, the plugin would also find http(s)://www.domain.tld/path/some-alias and would append the missing params automatically to the redirect url.
#### Summary of Changes
- Removed unnecessary comment blocks e.g.`$app = JFactory::getApplication(); or $uri = JUri::getInstance();`are common practice and should not need to be explained.
- "Reduced" the database access, refactored the insert / update methods and added separate try / catch blocks to support different error messages for both operations. I've also changed the error code in the catch block from 404 to 500 since in case of an error there, we're not dealing with a 404 anymore.
#### Testing Instructions
- Create a test article and a menu item for that article.
- Activate all SEF options in the global configuration and enable the redirect plugin in the plugin manager. Also activate the option to collect urls if it isn't already.
- If you click that link, everything should be good. Say, you named your article and the corresponding menu item "Test Article", leaving you with the following url: http://www.domain.tld/test-article
- Now, change /test-article to something that does not exist e.g. http://www.domain.tld/shenanigans. This should leave you with a classic 404 error page.
- Head over to the backend and check if that request was captured. If so, edit the entry and set the "Destination URL" to either http://www.domain.tld/test-article or just /test-article and enable the redirect.
- Visit http://www.domain.tld/shenanigans again. You should now be redirected to http://www.domain.tld/test-article.
- For the last part, visit the same url, but this time, add ?this=wontwork to the url so you end up with something similar to: http://www.domain.tld/shenanigans?this=wontwork
- Apply the patch and visit: http://www.domain.tld/shenanigans?this=willwork You should be redirected to your test article.
